nixos/modules/services/networking/cloudflared.nix

```diff
@@ -0,0 +1,95 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let cfg = config.services.cloudflared;
+in
+{
+  meta.maintainers = with maintainers; [ bbigras ];
+
+  options.services.cloudflared = {
+    enable = mkEnableOption "Cloudflared Argo Tunnel client daemon";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.cloudflared;
+      defaultText = "pkgs.cloudflared";
+      description = "The package to use for Cloudflared";
+    };
+
+    tunnels = mkOption {
+      description = ''
+        Cloudflare tunnels
+      '';
+      type = types.attrsOf (types.submodule ({ name, ... }: {
+        options = {
+          tunnel = mkOption {
+            type = with types; nullOr str;
+            default = null;
+            description = ''
+              Tunnel ID
+            '';
+            example = "00000000-0000-0000-0000-000000000000";
+          };
+
+          credentialsFile = mkOption {
+            type = with types; nullOr str;
+            default = null;
+            description = ''
+              credential file
+            '';
+          };
+
+          default = mkOption {
+            type = with types; nullOr str;
+            default = null;
+            description = ''
+              Catch-all if no ingress matches
+            '';
+            example = "http_status:404";
+          };
+
+          ingress = mkOption {
+            type = types.attrsOf types.str;
+            default = { };
+            example = {
+              "*.domain.com" = "http://localhost:80";
+              "*.anotherone.com" = "http://localhost:80";
+            };
+          };
+        };
+      }));
+
+      default = { };
+    };
+  };
+
+  config = {
+    systemd.services =
+      mapAttrs'
+        (name: tunnel:
+          let
+            fullConfig = {
+              tunnel = name;
+              "credentials-file" = tunnel.credentialsFile;
+              ingress = (map
+                (key: {
+                  hostname = key;
+                  service = getAttr key tunnel.ingress;
+                })
+                (attrNames tunnel.ingress)) ++ [{ service = tunnel.default; }];
+            };
+            mkConfigFile = pkgs.writeText "cloudflared.yml" (builtins.toJSON fullConfig);
+          in
+          nameValuePair "cloudflared-tunnel-${name}" ({
+            after = [ "network.target" ];
+            wantedBy = [ "multi-user.target" ];
+            serviceConfig = {
```

Comment:
I'm guessing running as `root` is a hard requirement?

Reply:
Maybe not, I'll try.

I'm guessing I can't use systemd `DynamicUser` with a secret file. I would need to create a `cloudflared` user, right?

Reply:
Ideally we could use `LoadCredential` to avoid the need for a dedicated user, but @happysalada has some less than happy experiences with that on NixOS atm... see https://github.com/systemd/systemd/issues/19604 for an ongoing discussion.